src/Distributed/partition_assemble.jl

```diff
@@ -21,11 +23,92 @@ and `arch`itecture.
 """
 function assemble(c_local::AbstractVector, Nc, Nr, r, arch) 
     c_global = zeros(eltype(c_local), Nc*Nr+1)
-    c_global[1 + (r-1) * Nc : Nc * r] .= c[1:end-1]
-    r == Nr && (c_global[end] = c[end])
+    c_global[1 + (r-1) * Nc : Nc * r] .= c_local[1:end-1]
+    r == Nr && (c_global[end] = c_local[end])
 
     MPI.Allreduce!(c_global, +, arch.communicator)
 
     return c_global
 end
 
+assemble(c::Tuple, Nc, Nr, r, arch) = (c[2] - r * (c[2] - c[1]), c[2] - (r - Nr) * (c[2] - c[1]))
+
+"""
+    partition_global_array(arch, c_global, (Nx, Ny, Nz))
+
+partition a global array (2D of size Nx, Ny or 3D of size Nx, Ny, Nz) in local arrays.
+Usefull for boundary arrays, forcings and initial conditions
+"""
+partition_global_array(arch, c_global::Function, N) = c_global 
+
+function partition_global_array(arch, c_global::AbstractArray, N) 
+    c_global = arch_array(CPU(), c_global)
+    Rx, Ry, Rz = R = arch.ranks
+    ri, rj, rk = r = arch.local_index
+
+    dims = length(size(c_global))
+
+    if dims == 2 
+        nx, ny = n = Int.(N[1:2] ./ R[1:2])
+
+        c_local = zeros(eltype(c_global), nx, ny)
+
+        c_local .= c_global[1 + (ri-1) * nx : nx * ri, 
+                            1 + (rj-1) * ny : ny * rj]
+    
+        return arch_array(child_architecture(arch), c_local)
+    else
+        nx, ny, nz = n = Int.(N ./ R)
+
+        c_local = zeros(eltype(c_global), nx, ny, nz)
+
+        c_local .= c_global[1 + (ri-1) * nx : nx * ri, 
+                            1 + (rj-1) * ny : ny * rj, 
+                            1 + (rk-1) * nz : nz * rk]
+
+        return arch_array(child_architecture(arch), c_local)
+    end
+end
+
+"""
+    reconstruct_global_array(arch, c_global, (nx, ny, nz))
+
+reconstruct local arrays (2D of size nx, ny or 3D of size nx, ny, nz) in local arrays.
+Usefull for boundary arrays, forcings and initial conditions
```

Comment:
```suggestion
Returns a local arrays (2D of size nx, ny or 3D of size nx, ny, nz).
Useful for boundary arrays, forcings and initial conditions.
```

I don't completely understand how this function is used. We are constructing a global array, but the input is `c_global`?

Reply:
Also is this just "construct" not "reconstruct"?